### PR TITLE
sql/spanner: Add integration test start

### PR DESCRIPTION
**Pull request by tmc**
This adds the basics of an integration testing workflow and the minimal necessary components of the spanner driver to support a basic integration test.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -563,4 +563,37 @@ jobs:
...
+  integration-spanner:
+    runs-on: ubuntu-latest
+    services:
+      spanner:
+        image: gcr.io/cloud-spanner-emulator/emulator:1.4.6
+        env:
+          SPANNER_EMULATOR_HOST: localhost:9010
+        ports:
+          - 9010:9010
+          - 9020:9020
```
Comment (a8m) [inline .github/workflows/ci.yml:576]: Why the second port is necessary?

File: internal/ci/main.go [hunk 2]
```diff
@@ -22,6 +22,8 @@ type Job struct {
 	Env     []string // env of service
 	Ports   []string // port mappings
 	Options []string // other options
+
```
Comment (a8m) [inline internal/ci/main.go:25]:
```suggestion
```

File: sql/spanner/convert.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	case *schema.FloatType:
+		f = t.T
+	case *schema.DecimalType:
+		f = t.T
+	case *schema.JSONType:
+		f = t.T
+	case *schema.SpatialType:
+		f = t.T
+	case *schema.BinaryType:
+		siz := ""
```
Comment (a8m) [inline sql/spanner/convert.go:38]:
```suggestion
		var size string
```

File: sql/spanner/convert.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+		return &schema.UnsupportedType{
+			T: c,
+		}, nil
+	}
+}
+
+func parseColumn(s string) (*columnDesc, error) {
+	cd := &columnDesc{}
+	// split up type into, base type, size, and other modifiers.
+	re := regexp.MustCompile(`(\w+)(?:\((-?\d+|MAX)\))?`)
```
Comment (a8m) [inline sql/spanner/convert.go:111]: compile it once. Move it to top-level declarations.

File: sql/spanner/convert.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+	m := re.FindStringSubmatch(strings.ToUpper(s))
+	if len(m) == 0 {
+		return nil, fmt.Errorf("parseColumn: invalid type: %q", s)
+	}
+	cd.typ = m[1]
+	if len(m) > 2 {
+		size, _ := strconv.Atoi(m[2])
```
Comment (a8m) [inline sql/spanner/convert.go:118]: Handle the error here?

File: sql/spanner/inspect.go [hunk 6]
```diff
@@ -501,8 +501,9 @@ type (
 
 	// columnDesc represents a column descriptor.
 	columnDesc struct {
-		typ  string
-		size int
+		typ       string
+		size      int
+		sizeIsMax bool
```
Comment (a8m) [inline sql/spanner/inspect.go:506]:
```suggestion
		maxSize bool
```

File: sql/spanner/migrate.go [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+func defaultValue(c *schema.Column) (string, error) {
+	switch x := c.Default.(type) {
+	case *schema.Literal:
+		return sqlx.SingleQuote(x.V)
+	case *schema.RawExpr:
+		return x.X, nil
+	default:
+		return "", fmt.Errorf("unexpected default value type: %T", x)
+	}
+}
```
Comment (a8m) [inline sql/spanner/migrate.go:315]: Unrelated to this change, but it feels like some migration parts can move to `sqlx`.

**Review by a8m — APPROVED** [review]
Looks good, minor comments.

After merging it, I think we/I can rebase the `spanner` branch.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (tmc) [reply .github/workflows/ci.yml:576]: It's not, this just mirrors what `gcloud emulators spanner start` sets up. We can pull the extra port.

**Review by a8m — APPROVED** [review]

**Pull request merged**